File: examples/python/oha/cmd/update.py
#!/usr/bin/python3


"""Command ``update`` to change already existing records within the db.

It can repeat the same operation in loop or a defined numbers of times. It can
be possible to define a delay between one operation and the next one and also
after how many operations commit the changes.
"""


import time
import click

import cx_Oracle


@click.command()
@click.option('--loop/--no-loop',
              type=bool,
              default=False,
              help="repeat the same operation forever")
@click.option('--iters',
              type=int,
              default=1,
              help='repeat the same operation a given number of times')
@click.option('--delay',
              type=float,
              default=0.25,
              help='time to wait before the next iteration')
@click.option('--commit-every',
              type=int,
              default=1,
              help='after how many operations perform a commit')
@click.pass_context
def update(ctx,
           loop: bool,
           iters: int,
           delay: float,
           commit_every: int) -> None:
    """Update records within the database"""

    # Define query parameters
    table = ctx.obj.conf["database"]["table"]
    args = {"DEPARTMENT_NAME": 0}
    conds = {"DEPARTMENT_ID": 1}

    iters = 0 if loop else iters
    step = 1
    try:
        while loop or step <= iters:
            # Prepare query with updated conditions
            sets = [f"{arg}='pippo{val + step}'" for arg, val in args.items()]
            wheres = [f"{arg}={val + step}" for arg, val in conds.items()]
            query = (f"UPDATE {table} "
                     f"SET {', '.join(sets)} "
                     f"WHERE {', '.join(wheres)}")

            # Execute query
            ctx.obj.cur.execute(query)
            click.echo(f"[{step}/{iters}] - {query}")

            # Commit changes
            if step % commit_every == 0:
                ctx.obj.conn.commit()
                click.echo(f"[{step}/{iters}] - COMMIT")

            step += 1
            time.sleep(delay)

        # Check the last commit
        if iters % commit_every != 0:
            ctx.obj.conn.commit()
            click.echo(f"[{iters}/{iters}] - COMMIT")
    except cx_Oracle.DatabaseError as err:
        click.echo(err)
        ctx.exit(1)
    except KeyboardInterrupt as _:
        click.echo("Error - Interrupted by the user")

File: apps/golang/oracle-ha/cmd/update.go
/*
Copyright © 2022 Andrea Lauri <andrea.lauri86@gmail.com>

Command ``update`` to change already existing records within the db.

It can repeat the same operation in loop or a defined numbers of times. It can
be possible to define a delay between one operation and the next one and also
after how many operations commit the changes.
*/
package cmd

import (
	"fmt"
	"time"

	"github.com/spf13/cobra"
	"github.com/spf13/viper"

	"github.com/alauri/oracle-ha-apps/oracle-ha/db"
)

// updateCmd represents the update command
var updateCmd = &cobra.Command{
	Use:   "update",
	Short: "Update records into the table.",
	Long:  `Update records into the table.`,
	Run: func(cmd *cobra.Command, args []string) {
		dsn, _ := cmd.Flags().GetInt("dsn")
		loop, _ := cmd.Flags().GetBool("loop")
		iters, _ := cmd.Flags().GetInt("iters")
		delay, _ := cmd.Flags().GetFloat64("delay")
		commit_every, _ := cmd.Flags().GetInt("commit-every")

		// Define query parameters
		table := viper.GetViper().GetString("database.table")
		conditions := Conds{1}
		arguments := Arguments{0}

		// Retrieve a fresh Database connection
		conn, err := db.GetDatabase(dsn)
		if err != nil {
			panic(err)
		}

		if loop {
			iters = 0
		}
		step := 1
		for {
			// Exit condition
			if !loop && step > iters {
				break
			}

			// Prepare query with updated conditions
			sets := Arguments{arguments.arg + step}
			wheres := Conds{conditions.id + step}
			query := fmt.Sprintf("UPDATE %s SET arg=%d WHERE id=%d",
				table, sets.arg, wheres.id)

			// Perform the query
			db.DoQuery(conn, query)
			fmt.Fprintln(cmd.OutOrStdout(),
				fmt.Sprintf("[%d/%d] - %s", step, iters, query))

			// Commit changes
			if step%commit_every == 0 {
				// TODO: insert a database commit operation here
				fmt.Fprintln(cmd.OutOrStdout(),
					fmt.Sprintf("[%d/%d] - COMMIT", step, iters))
			}

			step += 1
			time.Sleep(time.Duration(delay) * time.Second)
		}

		// Chech the last commit
		if iters%commit_every != 0 {
			// TODO: insert a database commit operation here
			fmt.Fprintln(cmd.OutOrStdout(),
				fmt.Sprintf("[%d/%d] - COMMIT", iters, iters))
		}
	},
}

func init() {
	updateCmd.Flags().Bool("loop", false, "repeat the same operation forever")
	updateCmd.Flags().Int("iters", 1, "repeat the same operation a given number of times")
	updateCmd.Flags().Float64("delay", 0.25, "time to wait before the next iteration")
	updateCmd.Flags().Int("commit-every", 1, "after how many operations perform a commit")
}
